### PR TITLE
Fix truncated backprop through time when set on LightningModule and not Trainer

### LINKED ISSUE #8802: Setting lightningmodule.truncated_bptt_steps>0 will not enable tBPTT
## 🐛 Bug
The _tbptt_split_batch function in training_batch_loop.py ignores the value lightningmodule.truncated_bptt_steps. Setting truncated_bptt_steps via Trainer will work fine but it is deprecated.

   ```
 def _tbptt_split_batch(self, batch: Any) -> List[Any]:
        """Splits a single batch into a list of sequence steps for tbptt.

        Args:
            batch: the current batch to split
        """
        splits = [batch]
        if self.trainer.truncated_bptt_steps is not None:
            model_ref = self.trainer.lightning_module
            with self.trainer.profiler.profile("tbptt_split_batch"):
                splits = model_ref.tbptt_split_batch(batch, self.trainer.truncated_bptt_steps)
        return splits
```

- PyTorch Lightning Version (e.g., 1.3.0): 1.4.1

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #8802

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 1]
```diff
@@ -451,11 +451,13 @@ def _tbptt_split_batch(self, batch: Any) -> List[Any]:
...
-            model_ref = self.trainer.lightning_module
-            with self.trainer.profiler.profile("tbptt_split_batch"):
-                splits = model_ref.tbptt_split_batch(batch, self.trainer.truncated_bptt_steps)
+        tbptt_steps = self._truncated_bptt_steps()
+        if tbptt_steps == 0:
+            return [batch]
+
+        model_ref = self.trainer.lightning_module
+        with self.trainer.profiler.profile("tbptt_split_batch"):
+            splits = model_ref.tbptt_split_batch(batch, tbptt_steps)
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:460]: when the trainer deprecation of the tbtt argument ends, do you think we should still pass the steps to the hook or should the module directly use its property?

**Review by awaelchli — APPROVED** [review]

File: tests/models/test_truncated_bptt.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+# Copyright The PyTorch Lightning team.
```
Comment (carmocca) [inline tests/models/test_truncated_bptt.py:1]: Since you created this file, can you also move the test `test_tbptt_log` in `tests/trainer/logging_/test_train_loop_logging.py`?

**Review by carmocca — APPROVED** [review]

File: tests/models/test_truncated_bptt.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    )
+    model.example_input_array = torch.randn(5, truncated_bptt_steps)
+
+    trainer_tbptt_steps = None if property_on_module else truncated_bptt_steps
+
+    # fit model
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        truncated_bptt_steps=trainer_tbptt_steps,
```
Comment (tchaton) [inline tests/models/test_truncated_bptt.py:92]:
```suggestion
        truncated_bptt_steps=8 if property_on_module else trainer_tbptt_steps, # the LightningModule has priority.
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/loops/batch/training_batch_loop.py:460]: I believe so. 

The step property today is serving 2 purposes:
- its presence & value indicates to the Trainer whether the user wants to use tbptt (ie change the loop behavior)
- its value is used for actually splitting batches

After the removal on the trainer side, one approach could be that we offer a TBPTTMixin, where `tbptt_split_batch` is a required method to implement. The user inherits this along with their LightningModule. 

```
class TBPTTModel(TBPTTMixin, LightningModule):
...
```

Then the Trainer could statically determine it should configure a corresponding tbptt_loop to run during training.

let's continue discussing on https://github.com/PyTorchLightning/pytorch-lightning/issues/8732 : )

File: tests/models/test_truncated_bptt.py (hunk 3 above)
Comment (ananthsub) [reply tests/models/test_truncated_bptt.py:92]: @tchaton I wonder if we should raise a misconfiguration error instead if the trainer and lightning module both set values for this and they're not the same

Comment (ananthsub) [reply tests/models/test_truncated_bptt.py:92]: i will send out a separate PR for this
